### ISSUE TITLE
Very slow push-pull

### ISSUE THREAD

**Issue by dyoniz** [zeromq/netmq#556]
TaskVent + TaskWork + TaskSink from [ZeroMQ Guide](http://zguide.zeromq.org/page:all#Divide-and-Conquer) implemented using NetMQ, work super slow!
Output from TaskVent: Total expected cost: 5605 ms
Output from TaskSink: Total elepsed time: 6415 ms (delay is over 800 ms!)
Same results with IPC.
Same example, implemented with clrzmq works perfectly fine:
TaskVent: Total expected cost: 5605 ms
TaskSink: Total elapsed time: 5636 ms (delay is only 31 ms)
If you replace TaskWork or TaskSink (or both) with one implemented with clrzmq, then everything works fine (delay is less than 50 ms).

Maybe it is a AsyncIO bug.
NetMQ version is 3.3.3.1 and AsyncIO version is 0.1.18.0.

``` csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NetMQ;
using NetMQ.Sockets;

namespace LearnNetMQ
{
    static partial class Program
    {
        static void Main(string[] args)
        {
            if (args.Length < 1)
            {
                Console.WriteLine("Usage: .\\{0} ExampleName [ExampleArgs]", AppDomain.CurrentDomain.FriendlyName);
                return;
            }

            try
            {
                typeof(Program).GetMethod(args[0]).Invoke(null, new object[] { args.SubArray(1) });
            }
            catch (MissingMethodException)
            {
                Console.WriteLine("Unknown example");
            }
        }

        public static void TaskVent(string[] args)
        {
            //
            // Task ventilator
            // Binds PUSH socket to tcp://localhost:5557
            // Sends batch of tasks to workers via that socket
            //

            // Bind ventilator
            // Connect to sink

            //using (var sender = new PushSocket("@ipc://TaskVent"))
            //using (var sink = new PushSocket(">ipc://TaskSink"))
            using (var sender = new PushSocket("@tcp://*:5557"))
            using (var sink = new PushSocket(">tcp://127.0.0.1:5558"))
            {
                Console.WriteLine("Press ENTER when the workers are ready...");
                Console.ReadKey(true);
                Console.WriteLine("Sending tasks to workers...");

                // The first message is "0" and signals start of batch
                sink.SendFrame(new byte[] { 0x00 });

                var rnd = new Random(22);

                // Send 100 tasks
                int i = 0;
                long total_msec = 0; // Total expected cost in msecs
                for (; i < 100; i++)
                {
                    // Random workload from 1 to 100 msecs
                    int workload = rnd.Next(100) + 1;
                    total_msec += workload;
                    Console.WriteLine(workload);
                    sender.SendFrame(BitConverter.GetBytes(workload));
                }
                Console.WriteLine("Total expected cost: {0} ms", total_msec);
            }
        }

        public static void TaskWork(string[] args)
        {
            //
            // Task worker
            // Connects PULL socket to tcp://localhost:5557
            // Collects workloads from ventilator via that socket
            // Connects PUSH socket to tcp://localhost:5558
            // Sends results to sink via that socket
            //

            // Create and connect socket to receive tasks from ventilator
            // Create and connect socket to send results to sink
            Console.WriteLine("Connecting to ventilator and sink...");

            //using (var receiver = new PullSocket(">ipc://TaskVent"))
            //using (var sink = new PushSocket(">ipc://TaskSink"))
            using (var receiver = new PullSocket(">tcp://127.0.0.1:5557"))
            using (var sink = new PushSocket(">tcp://127.0.0.1:5558"))
            {
                // Process tasks forever
                Console.WriteLine("Waiting for tasks...");
                while (true)
                {
                    byte[] frame = receiver.ReceiveFrameBytes();
                    int workload = BitConverter.ToInt32(frame, 0);
                    Console.WriteLine("{0}.", workload);

                    // Do the work
                    Thread.Sleep(workload);

                    // Push results to sink
                    sink.SendFrameEmpty();
                }
            }
        }

        public static void TaskSink(string[] args)
        {
            //
            // Task sink
            // Binds PULL socket to tcp://localhost:5558
            // Collects results from workers via that socket
            //

            // Create and bind our sink socket
            Console.WriteLine("Binding sink to port 5558...");
            //using (var sink = new PullSocket("@ipc://TaskSink"))
            using (var sink = new PullSocket("@tcp://*:5558"))
            {
                // Wait for start of batch
                Console.WriteLine("Waiting for the start signal...");
                sink.ReceiveFrameBytes();

                Console.WriteLine("Processing results...");
                // Start our clock now
                var stopwatch = new Stopwatch();
                stopwatch.Start();

                // Process 100 confirmations
                for (int i = 0; i < 100; i++)
                {
                    sink.ReceiveFrameBytes();

                    if ((i / 10) * 10 == i)
                        Console.Write(":");
                    else
                        Console.Write(".");
                }

                stopwatch.Stop();
                Console.WriteLine();
                Console.WriteLine("Total elepsed time: {0} ms", stopwatch.ElapsedMilliseconds);
            }
        }

        public static T[] SubArray<T>(this T[] arr, int startIndex)
        {
            if (arr == null)
                return null;
            int length = arr.Length - startIndex;
            if (length <= 0)
                return new T[0];
            T[] result = new T[length];
            Array.Copy(arr, startIndex, result, 0, length);
            return result;
        }
    }
}
```

**Comment by drewnoakes** [comment]
Off the top of my head: Are you sure you're not measuring JIT and warmup
time? Might be worth sending a few messages through before commencing the
timed run.

**Comment by dyoniz** [comment]
Ok, today I tried to send 500 tasks instead of 100 (using single TaskWorker as before) and got these results:
TaskVent: Total expected cost: 26489 ms
TaskSink: Total elapsed time: 26524 ms (35 ms delay)
Thats weird.
So I tried to send 100 tasks again and got these results:
Total expected cost: 5605 ms
Total elapsed time: 5620 ms (15 ms delay)
So now everythink works fine. And I didn't make any changes to the code since yesterday. So I asume that was something on my system, that caused the delay. But whatever that was, the exact same code with NetMQ sockets replaced with clrzmq sockets worked perfectly fine. Both programs were compiled using .NET Framework 4.0 and ran on Windows 7 x64.
I gues It has something to do with AsyncIO.
